File: cslavb/Csla/Server/Hosts/Silverlight/WcfPortal.vb
Imports System
Imports System.Configuration
Imports Csla.Serialization.Mobile
Imports System.ServiceModel
Imports System.ServiceModel.Activation
Imports Csla.Core
Imports System.Security.Principal
Imports Csla.Properties

Namespace Server.Hosts.Silverlight

    ''' <summary>
    ''' Exposes server-side DataPortal functionality
    ''' through WCF.
    ''' </summary>
    ''' <remarks></remarks>
    <AspNetCompatibilityRequirements(RequirementsMode = AspNetCompatibilityRequirementsMode.Allowed)> _
    Public Class WcfPortal
        Implements IWcfPortal

#Region "Factory Loader"

        Private Shared _factoryLoader As IMobileFactoryLoader

        Public Shared Property FactoryLoader() As IMobileFactoryLoader
            Get
                If _factoryLoader Is Nothing Then
                    Dim setting As String = ConfigurationManager.AppSettings("CslaMobileFactoryLoader")

                    If String.IsNullOrEmpty(setting) Then
                        _factoryLoader = CType(Activator.CreateInstance(Type.GetType(setting, True, True)), IMobileFactoryLoader)
                    Else
                        _factoryLoader = New MobileFactoryLoader()
                    End If
                End If

                Return _factoryLoader
            End Get
            Set(ByVal value As IMobileFactoryLoader)
                _factoryLoader = value
            End Set
        End Property

#End Region

#Region "IWcfPortal Members"

        ''' <summary>
        ''' Create a new business object.
        ''' </summary>
        ''' <param name="request">The request parameter object.</param>
        Public Function Create(ByVal request As CriteriaRequest) As WcfResponse Implements IWcfPortal.Create
            Dim result As WcfResponse = New WcfResponse()

            Try
                request = ConvertRequest(request)
                'unpack criteria data into object
                Dim criteria As Object = GetCriteria(request.CriteriaData)

                'load type for business object
                Dim t = Type.GetType(request.TypeName)
                If t Is Nothing Then
                    Throw New InvalidOperationException(String.Format(Resources.ObjectTypeCouldNotBeLoaded, request.TypeName))
                End If

                SetContext(request)
                Dim o As Object = Nothing
                Dim factoryInfo = GetMobileFactoryAttribute(t)
                If factoryInfo Is Nothing Then
                    If criteria IsNot Nothing Then
                        o = Csla.DataPortal.Create(criteria)
                    Else
                        o = Csla.DataPortal.Create(t)
                    End If
                Else

                    If String.IsNullOrEmpty(factoryInfo.CreateMethodName) Then
                        Throw New InvalidOperationException(Resources.CreateMethodNameNotSpecified)
                    End If

                    Dim f As Object = FactoryLoader.GetFactory(factoryInfo.FactoryTypeName)
                    If criteria IsNot Nothing Then
                        o = Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.CreateMethodName, criteria)
                    Else
                        o = Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.CreateMethodName)
                    End If
                End If

                result.ObjectData = MobileFormatter.Serialize(o)
                result.GlobalContext = MobileFormatter.Serialize(ApplicationContext.GlobalContext)

            Catch ex As Csla.Reflection.CallMethodException
                result.ErrorData = New WcfErrorInfo(ex.InnerException)
            Catch ex As Exception
                result.ErrorData = New WcfErrorInfo(ex)
            Finally
                ClearContext()
            End Try

            Return ConvertResponse(result)

        End Function

        ''' <summary>
        ''' Get an existing business object.
        ''' </summary>
        ''' <param name="request">The request parameter object.</param>
        Public Function Fetch(ByVal request As CriteriaRequest) As WcfResponse Implements IWcfPortal.Fetch
            Dim result As WcfResponse = New WcfResponse()

            Try
                request = ConvertRequest(request)
                'unpack criteria data into object
                Dim criteria As Object = GetCriteria(request.CriteriaData)

                'load type for business object
                Dim t = Type.GetType(request.TypeName)
                If t Is Nothing Then
                    Throw New InvalidOperationException( _
                        String.Format(Resources.ObjectTypeCouldNotBeLoaded, request.TypeName))
                End If

                SetContext(request)
                Dim o As Object = Nothing
                Dim factoryInfo = GetMobileFactoryAttribute(t)
                If factoryInfo Is Nothing Then
                    If criteria Is Nothing Then
                        o = Csla.DataPortal.Fetch(t)
                    Else
                        o = Csla.DataPortal.Fetch(t, criteria)
                    End If
                Else

                    If String.IsNullOrEmpty(factoryInfo.FetchMethodName) Then
                        Throw New InvalidOperationException(Resources.FetchMethodNameNotSpecified)
                    End If

                    Dim f As Object = FactoryLoader.GetFactory(factoryInfo.FactoryTypeName)
                    If criteria IsNot Nothing Then
                        o = Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.FetchMethodName, criteria)
                    Else
                        o = Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.FetchMethodName)
                    End If
                End If

                result.ObjectData = MobileFormatter.Serialize(o)
                result.GlobalContext = MobileFormatter.Serialize(ApplicationContext.GlobalContext)
            Catch ex As Csla.Reflection.CallMethodException
                result.ErrorData = New WcfErrorInfo(ex.InnerException)
            Catch ex As Exception
                result.ErrorData = New WcfErrorInfo(ex)
            Finally
                ClearContext()
            End Try

            Return ConvertResponse(result)

        End Function

        ''' <summary>
        ''' Update a business object
        ''' </summary>
        ''' <param name="request">The request parameter object.</param>
        Public Function Update(ByVal request As UpdateRequest) As WcfResponse Implements IWcfPortal.Update
            Dim result As WcfResponse = New WcfResponse()

            Try
                request = ConvertRequest(request)
                'unpack object
                Dim obj As Object = GetCriteria(request.ObjectData)

                'load type for business object
                Dim t = obj.GetType()

                Dim o As Object = Nothing
                Dim factoryInfo = GetMobileFactoryAttribute(t)
                If factoryInfo Is Nothing Then
                    SetContext(request)
                    o = Csla.DataPortal.Update(obj)
                Else
                    If String.IsNullOrEmpty(String.IsNullOrEmpty(factoryInfo.UpdateMethodName)) Then
                        Throw New InvalidOperationException(Resources.UpdateMethodNameNotSpecified)
                    End If

                    SetContext(request)
                    Dim f As Object = FactoryLoader.GetFactory(factoryInfo.FactoryTypeName)
                    o = Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.UpdateMethodName, obj)

                End If

                result.ObjectData = MobileFormatter.Serialize(o)
                result.GlobalContext = MobileFormatter.Serialize(ApplicationContext.GlobalContext

            Catch ex As Exception
                result.ErrorData = New WcfErrorInfo(ex)
            Finally
                ClearContext()
            End Try

            Return ConvertResponse(result)
        End Function

        ''' <summary>
        ''' Delete a business object.
        ''' </summary>
        ''' <param name="request">The request parameter object.</param>
        Public Function Delete(ByVal request As CriteriaRequest) As WcfResponse Implements IWcfPortal.Delete
            Dim result As WcfResponse = New WcfResponse()

            Try
                request = ConvertRequest(request)
                'unpack criteria data into object
                Dim criteria As Object = GetCriteria(request.CriteriaData)

                'load type for business object
                Dim t = Type.GetType(request.TypeName)
                If t Is Nothing Then
                    Throw New InvalidOperationException( _
                        String.Format(Resources.ObjectTypeCouldNotBeLoaded, request.TypeName))
                End If

                SetContext(request)
                Dim factoryInfo = GetMobileFactoryAttribute(t)
                If factoryInfo Is Nothing Then
                    Csla.DataPortal.Delete(criteria)
                Else
                    If String.IsNullOrEmpty(factoryInfo.DeleteMethodName) Then
                        Throw New InvalidOperationException(Resources.DeleteMethodNameNotSpecified)
                    End If

                    Dim f As Object = FactoryLoader.GetFactory(factoryInfo.FactoryTypeName)

                    If criteria IsNot Nothing Then
                        Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.DeleteMethodName, criteria)
                    Else
                        Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.DeleteMethodName)
                    End If
                End If

                result.GlobalContext = MobileFormatter.Serialize(ApplicationContext.GlobalContext)

            Catch ex As Csla.Reflection.CallMethodException
                result.ErrorData = New WcfErrorInfo(ex.InnerException)
            Catch ex As Exception
                result.ErrorData = New WcfErrorInfo(ex)
            Finally
                ClearContext()
            End Try

        End Function

#End Region

#Region "Context and Criteria"

        Private Sub SetContext(ByVal request As CriteriaRequest)
            ApplicationContext.SetContext(CType(MobileFormatter.Deserialize(request.ClientContext), ContextDictionary), CType(MobileFormatter.Deserialize(request.GlobalContext), ContextDictionary))
            ApplicationContext.User = CType(MobileFormatter.Deserialize(request.Principal), IPrincipal)
        End Sub

        Private Sub SetContext(ByVal request As UpdateRequest)
            ApplicationContext.SetContext(CType(MobileFormatter.Deserialize(request.ClientContext), ContextDictionary), CType(MobileFormatter.Deserialize(request.GlobalContext), ContextDictionary))
            ApplicationContext.User = CType(MobileFormatter.Deserialize(request.Principal), IPrincipal)
        End Sub

        Private Shared Sub ClearContext(ByVal request As UpdateRequest)
            ApplicationContext.Clear()
            If ApplicationContext.AuthenticationType <> "Windows" Then
                ApplicationContext.User = New System.Security.Principal.GenericPrincipal(New System.Security.Principal.GenericIdentity(String.Empty), New String() {})
            End If
        End Sub

        Private Shared Function GetCriteria(ByVal criteriaData() As Byte) As Object
            Dim criteria As Object = Nothing
            If criteriaData IsNot Nothing Then
                criteria = MobileFormatter.Deserialize(criteriaData)
            End If
            Return criteria
        End Function

#End Region

#Region "Mobile Factory"

        Private Shared Function GetMobileFactoryAttribute(ByVal objectType As Type) As Csla.Silverlight.MobileFactoryAttribute
            Dim result = objectType.GetCustomAttributes(GetType(Csla.Silverlight.MobileFactoryAttribute), True)

            If result IsNot Nothing AndAlso result.Length > 0 Then
                Return CType(result(0), Csla.Silverlight.MobileFactoryAttribute)
            Else
                Return Nothing
            End If

        End Function

#End Region

#Region "Extention Method for Requests"

        ''' <summary>
        ''' Override to convert the request data before it
        ''' is transferred over the network.
        ''' </summary>
        ''' <param name="request">Request object.</param>
        Protected Overridable Function ConvertRequest(ByVal request As UpdateRequest) As UpdateRequest
            Return request
        End Function

        ''' <summary>
        ''' Override to convert the request data before it
        ''' is transferred over the network.
        ''' </summary>
        ''' <param name="request">Request object.</param>
        Protected Overridable Function ConvertRequest(ByVal request As CriteriaRequest) As CriteriaRequest
            Return request
        End Function

        ''' <summary>
        ''' Override to convert the response data after it
        ''' comes back from the network.
        ''' </summary>
        ''' <param name="response">Response object.</param>
        Protected Overridable Function ConvertResponse(ByVal response As WcfResponse) As WcfResponse
            Return response
        End Function

#End Region

    End Class
End Namespace

File: Source/Csla.core/Server/Hosts/Silverlight/WcfPortal.cs
using System;
using System.Configuration;
using Csla.Serialization.Mobile;
using System.ServiceModel;
using System.ServiceModel.Activation;
using Csla.Core;
using System.Security.Principal;
using Csla.Properties;

namespace Csla.Server.Hosts.Silverlight
{
  /// <summary>
  /// Exposes server-side DataPortal functionality
  /// through WCF.
  /// </summary>
  [AspNetCompatibilityRequirements(RequirementsMode = AspNetCompatibilityRequirementsMode.Allowed)]
   public class WcfPortal : IWcfPortal
  {
    #region Factory Loader

    private static IMobileFactoryLoader _factoryLoader;
    /// <summary>
    /// Gets or sets a delegate reference to the method
    /// called to create instances of factory objects
    /// as requested by the MobileFactory attribute on
    /// a CSLA Light business object.
    /// </summary>
    public static IMobileFactoryLoader FactoryLoader 
    {
      get
      {
        if (_factoryLoader == null)
        {
          string setting = ConfigurationManager.AppSettings["CslaMobileFactoryLoader"];
          if (!string.IsNullOrEmpty(setting))
            _factoryLoader = 
              (IMobileFactoryLoader)Activator.CreateInstance(Type.GetType(setting, true, true));
          else
            _factoryLoader = new MobileFactoryLoader();
        }
        return _factoryLoader;
      }
      set
      {
        _factoryLoader = value;
      }
    }

    #endregion

    #region IWcfPortal Members

    /// <summary>
    /// Create a new business object.
    /// </summary>
    /// <param name="request">The request parameter object.</param>
    public WcfResponse Create(CriteriaRequest request)
    {
      var result = new WcfResponse();
      try
      {
        request = ConvertRequest(request);
        // unpack criteria data into object
        object criteria = GetCriteria(request.CriteriaData);

        // load type for business object
        var t = Type.GetType(request.TypeName);
        if (t == null)
          throw new InvalidOperationException(
            string.Format(Resources.ObjectTypeCouldNotBeLoaded, request.TypeName));

        SetContext(request);
        object o = null;
        var factoryInfo = GetMobileFactoryAttribute(t);
        if (factoryInfo == null)
        {
          if (criteria != null)
          {
            o = Csla.DataPortal.Create(t, criteria);
          }
          else
          {
            o = Csla.DataPortal.Create(t);
          }
        }
        else
        {
          if (string.IsNullOrEmpty(factoryInfo.CreateMethodName))
            throw new InvalidOperationException(Resources.CreateMethodNameNotSpecified);

          object f = FactoryLoader.GetFactory(factoryInfo.FactoryTypeName);
          if (criteria != null)
            o = Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.CreateMethodName, criteria);
          else
            o = Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.CreateMethodName);
        }
        result.ObjectData = MobileFormatter.Serialize(o);
        result.GlobalContext = MobileFormatter.Serialize(ApplicationContext.GlobalContext);
      }
      catch (Csla.Reflection.CallMethodException ex)
      {
        result.ErrorData = new WcfErrorInfo(ex.InnerException);
      }
      catch (Exception ex)
      {
        result.ErrorData = new WcfErrorInfo(ex);
      }
      finally
      {
        ClearContext();
      }
      return ConvertResponse(result);
    }

    /// <summary>
    /// Get an existing business object.
    /// </summary>
    /// <param name="request">The request parameter object.</param>
    public WcfResponse Fetch(CriteriaRequest request)
    {
      var result = new WcfResponse();
      try
      {
        request = ConvertRequest(request);
        // unpack criteria data into object
        object criteria = GetCriteria(request.CriteriaData);

        // load type for business object
        var t = Type.GetType(request.TypeName);
        if (t == null)
          throw new InvalidOperationException(
            string.Format(Resources.ObjectTypeCouldNotBeLoaded, request.TypeName));
        SetContext(request);
        object o = null;
        var factoryInfo = GetMobileFactoryAttribute(t);
        if (factoryInfo == null)
        {
          if (criteria == null)
            o = Csla.DataPortal.Fetch(t);
          else
            o = Csla.DataPortal.Fetch(t, criteria);
        }
        else
        {
          if (string.IsNullOrEmpty(factoryInfo.FetchMethodName))
            throw new InvalidOperationException(Resources.FetchMethodNameNotSpecified);

          object f = FactoryLoader.GetFactory(factoryInfo.FactoryTypeName);
          if (criteria != null)
            o = Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.FetchMethodName, criteria);
          else
            o = Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.FetchMethodName);
        }
        result.ObjectData = MobileFormatter.Serialize(o);
        result.GlobalContext = MobileFormatter.Serialize(ApplicationContext.GlobalContext);
      }
      catch (Csla.Reflection.CallMethodException ex)
      {
        result.ErrorData = new WcfErrorInfo(ex.InnerException);
      }
      catch (Exception ex)
      {
        result.ErrorData = new WcfErrorInfo(ex);
      }
      finally
      {
        ClearContext();
      }
      return ConvertResponse(result);
    }

    /// <summary>
    /// Update a business object.
    /// </summary>
    /// <param name="request">The request parameter object.</param>
    public WcfResponse Update(UpdateRequest request)
    {
      var result = new WcfResponse();
      try
      {
        request = ConvertRequest(request);
        // unpack object
        object obj = GetCriteria(request.ObjectData);

        // load type for business object
        var t = obj.GetType();

        object o = null;
        var factoryInfo = GetMobileFactoryAttribute(t);
        if (factoryInfo == null)
        {
          SetContext(request);
          o = Csla.DataPortal.Update(obj);
        }
        else
        {
          string methodName;
          if (obj is Core.ICommandObject)
            methodName = factoryInfo.ExecuteMethodName;
          else
          {
            methodName = factoryInfo.UpdateMethodName;
          }
          if (string.IsNullOrEmpty(methodName))
            throw new InvalidOperationException(Resources.UpdateMethodNameNotSpecified);

          SetContext(request);
          object f = FactoryLoader.GetFactory(factoryInfo.FactoryTypeName);
            o = Csla.Reflection.MethodCaller.CallMethod(f, methodName, obj);
        }
        result.ObjectData = MobileFormatter.Serialize(o);
        result.GlobalContext = MobileFormatter.Serialize(ApplicationContext.GlobalContext);
      }
      catch (Csla.Reflection.CallMethodException ex)
      {
        result.ErrorData = new WcfErrorInfo(ex.InnerException);
      }
      catch (Exception ex)
      {
        result.ErrorData = new WcfErrorInfo(ex);
      }
      finally
      {
        ClearContext();
      }
      return ConvertResponse(result);
    }

    /// <summary>
    /// Delete a business object.
    /// </summary>
    /// <param name="request">The request parameter object.</param>
    public WcfResponse Delete(CriteriaRequest request)
    {
      var result = new WcfResponse();
      try
      {
        request = ConvertRequest(request);
        // unpack criteria data into object
        object criteria = GetCriteria(request.CriteriaData);

        // load type for business object
        var t = Type.GetType(request.TypeName);
        if (t == null)
          throw new InvalidOperationException(
            string.Format(Resources.ObjectTypeCouldNotBeLoaded, request.TypeName));

        SetContext(request);

        var factoryInfo = GetMobileFactoryAttribute(t);
        if (factoryInfo == null)
        {
          Csla.DataPortal.Delete(criteria);
        }
        else
        {
          if (string.IsNullOrEmpty(factoryInfo.DeleteMethodName))
            throw new InvalidOperationException(Resources.DeleteMethodNameNotSpecified);

          object f = FactoryLoader.GetFactory(factoryInfo.FactoryTypeName);
          if (criteria != null)
            Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.DeleteMethodName, criteria);
          else
            Csla.Reflection.MethodCaller.CallMethod(f, factoryInfo.DeleteMethodName);
        }
        result.GlobalContext = MobileFormatter.Serialize(ApplicationContext.GlobalContext);
      }
      catch (Csla.Reflection.CallMethodException ex)
      {
        result.ErrorData = new WcfErrorInfo(ex.InnerException);
      }
      catch (Exception ex)
      {
        result.ErrorData = new WcfErrorInfo(ex);
      }
      finally
      {
        ClearContext();
      }
      return ConvertResponse(result);
    }

    #endregion

    #region Context and Criteria
    private void SetContext(CriteriaRequest request)
    {
      ApplicationContext.SetContext((ContextDictionary)MobileFormatter.Deserialize(request.ClientContext), (ContextDictionary)MobileFormatter.Deserialize(request.GlobalContext));
      ApplicationContext.User = (IPrincipal)MobileFormatter.Deserialize(request.Principal);
      // set the thread's culture to match the client
      if (!string.IsNullOrEmpty(request.ClientCulture))
        System.Threading.Thread.CurrentThread.CurrentCulture =
          new System.Globalization.CultureInfo(request.ClientCulture);
      if (!string.IsNullOrEmpty(request.ClientUICulture))
        System.Threading.Thread.CurrentThread.CurrentUICulture =
          new System.Globalization.CultureInfo(request.ClientUICulture);
    }

    private void SetContext(UpdateRequest request)
    {
      ApplicationContext.SetContext((ContextDictionary)MobileFormatter.Deserialize(request.ClientContext), (ContextDictionary)MobileFormatter.Deserialize(request.GlobalContext));
      ApplicationContext.User = (IPrincipal)MobileFormatter.Deserialize(request.Principal);
      // set the thread's culture to match the client
      if (!string.IsNullOrEmpty(request.ClientCulture))
        System.Threading.Thread.CurrentThread.CurrentCulture =
          new System.Globalization.CultureInfo(request.ClientCulture);
      if (!string.IsNullOrEmpty(request.ClientUICulture))
        System.Threading.Thread.CurrentThread.CurrentUICulture =
          new System.Globalization.CultureInfo(request.ClientUICulture);
    }

    private static void ClearContext()
    {
      ApplicationContext.Clear();
      if (ApplicationContext.AuthenticationType != "Windows")
        ApplicationContext.User = new System.Security.Principal.GenericPrincipal(new System.Security.Principal.GenericIdentity(string.Empty),new string[]{});
    }

    private static object GetCriteria(byte[] criteriaData)
    {
      object criteria = null;
      if (criteriaData != null)
        criteria = MobileFormatter.Deserialize(criteriaData);
      return criteria;
    }
    #endregion

    #region Mobile Factory
    private static MobileFactoryAttribute GetMobileFactoryAttribute(Type objectType)
    {
      var result = objectType.GetCustomAttributes(typeof(MobileFactoryAttribute), true);
      if (result != null && result.Length > 0)
        return result[0] as MobileFactoryAttribute;
      else
        return null;
    }
    #endregion

    #region Extention Method for Requests

    /// <summary>
    /// Override to convert the request data before it
    /// is transferred over the network.
    /// </summary>
    /// <param name="request">Request object.</param>
    protected virtual UpdateRequest ConvertRequest(UpdateRequest request)
    {
      return request;
    }

    /// <summary>
    /// Override to convert the request data before it
    /// is transferred over the network.
    /// </summary>
    /// <param name="request">Request object.</param>
    protected virtual CriteriaRequest ConvertRequest(CriteriaRequest request)
    {
      return request;
    }

    /// <summary>
    /// Override to convert the response data after it
    /// comes back from the network.
    /// </summary>
    /// <param name="response">Response object.</param>
    protected virtual WcfResponse ConvertResponse(WcfResponse response)
    {
      return response;
    }

    #endregion
  }
}